### PR TITLE
[3/n] tls: add outbound tls dialer

### PR DESCRIPTION
**Pull request by jronak**
Add TLS dialer which will be used by the outbounds of all the transports. In the next pr, TODOs will be addressed.

### CODE REVIEW

File: transport/internal/tls/dialer/dialer_test.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+				Config: clientTLSConfig,
+			}
+			calledCustomDialer := false
+			if tt.withCustomDialer {
+				params.Dialer = func(ctx context.Context, network, address string) (net.Conn, error) {
+					calledCustomDialer = true
+					return (&net.Dialer{}).DialContext(ctx, network, address)
+				}
+			}
+			dialer := NewTLSDialer(params)
```
Comment (AllenLuUber) [inline transport/internal/tls/dialer/dialer_test.go:59]: can you move this line before it is actually used? L80. It got me confused a bit in the first pass that they we are setting a dialer before we listen.

File: transport/internal/tls/dialer/dialer_test.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+			conn, err := dialer.DialContext(context.Background(), "tcp", lis.Addr().String())
+			require.NoError(t, err)
+
+			_, ok := conn.(*tls.Conn)
+			assert.True(t, ok)
+
+			n, err := conn.Write([]byte(tt.data))
+			require.NoError(t, err)
+			assert.Equal(t, len(tt.data), n)
```
Comment (AllenLuUber) [inline transport/internal/tls/dialer/dialer_test.go:88]: `assert.Len(t, n, tt.data)`

File: transport/internal/tls/dialer/dialer_test.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+			assert.Equal(t, tt.data, string(buf))
+
+			if tt.withCustomDialer {
+				assert.True(t, calledCustomDialer)
+			}
+		})
+	}
+}
+
+func tlsConfigs(t *testing.T) (serverConfig *tls.Config, clientConfig *tls.Config) {
```
Comment (AllenLuUber) [inline transport/internal/tls/dialer/dialer_test.go:102]: oof I wish tls package provides some testing primitives so you don't have to do those.

Comment (jronak) [reply transport/internal/tls/dialer/dialer_test.go:102]: Yea I can move this method to tls/testscenario in a follow up

**Review by AllenLuUber — APPROVED** [review]

File: transport/internal/tls/dialer/dialer_test.go (hunk 3 above)
Comment (jronak) [reply transport/internal/tls/dialer/dialer_test.go:103]: #2176 done

**Pull request merged**